### DISCUSSION TITLE
How can I Monkey patch a request before fetching it?

### DISCUSSION THREAD

**Discussion by royhersh** [mswjs/msw#1405] [Q&A]
#### tl;dr:
I want to be able to monkey patch requests.
In older versions, I was able to override the request body, like the code example below.
I have upgrade msw to the latest and now the body is read-only and this breaks.

**Any idea how can I do that now ?**
 
```javascript
  rest.patch(
    '*/construction/rfis-qa/v3/projects/:projectId/rfi-types/:rfiTypeId',
    async (req, res, ctx) => {
      reqToV2(req); // converts the request url from '/v3/' to '/v2'
      req.body = adapters.path.v3.to.v2(req.body as V3);

      const originalResponse = await ctx.fetch(req);
      const originalResponseData: Response_V2 = await originalResponse.json();

      const responseV3: Response_V3 = adapters.patch.v2.to.v3(originalResponseData);

      return res(ctx.json(responseV3));
    },
  ),
```

#### Context

Our backend is going to switch to V3, which is the same data as V2 but with a different structure.
V3 doesn't exist yet, I want to mock V3 with msw from real V2 endpoints.
The illustration below explains what I'm trying to do
```mermaid
sequenceDiagram
    FrontEnd->>MSW: Req:V3
    MSW->>MSW:Converts V3 req body to V2
    MSW->>Backend: Req:V2
    Backend->>MSW: Res:V2
    MSW->>MSW:Converts V2 res to V3
    MSW->>FrontEnd: Res:V3
```

**Comment by kettanaito** [comment]
Hey, @royhersh.

As we've recently reworked how we store requests, the request body is indeed readonly now. This gives you the same behavior as the `Request` instance in the browser. 

To perform response patching now, you'd have to construct a new `Request` with the body you need. You can then proxy that request to the original (or different) URL using `ctx.fetch()` just as you do in your example. 

```ts
rest.patch(
  '*/construction/rfis-qa/v3/projects/:projectId/rfi-types/:rfiTypeId',
  async (req, res, ctx) => {
    const proxyReq = new Request(toV2Url(req.url), {
      ...req,
      body: adapters.path.v2.to.v2(req.body)
    })

    const originalResponse = await ctx.fetch(proxyReq);
    const originalResponseData: Response_V2 = await originalResponse.json();
    const responseV3: Response_V3 = adapters.patch.v2.to.v3(originalResponseData);

    return res(ctx.json(responseV3));
  },
),
```

The key points here is:

1. `proxyReq` which is a `new Request()`.
2. Create a proxy URl via `toV2URL()` giving it the original intercepted `req.url`.
3. Create a different body by assigning the `body` property in the `RequestInit`. 

The important part here is that you **mustn't mutate the original `req`**.

Let me know if this works.

**Reply by royhersh** [reply]
Hey @kettanaito, thanks for the fast reply!.

I tried to implement what you suggested, keeping in mind the key points.
It didn't work :-( 
On the line  
```javascript
const originalResponse = await ctx.fetch(proxyReq);
``` 
I get the error: `TypeError: Failed to execute 'fetch' on 'Window': Request with GET/HEAD method cannot have body.`
Which is weird, because I `console.log(proxyReq)` right before the fetch and `proxyReq.method === 'PATCH'`, it's also have empty Headers.

I think it may be related to that `req` and `proxyReq` has different type, WDYT?

**Reply by kettanaito** [reply]
Looks like the constructed `proxyReq` doesn't infer the intercepted request's method. Try setting it explicitly:

```js
const proxyReq = new Request(url, {
  method: req.method,
  body: '...'
})
```

**Reply by royhersh** [reply]
I tried that as well with no result.
Is there a way to use the same constructor of the the mocked request so they will have the same type?

**Reply by royhersh** [reply]
Ok, I think I figured it out.
There are two issues here:

1. `ctx.fetch` accept `MockedRequest` and not `Request`, so we better use `MockedRequest` instead.
2. With `MockedRequest`, the `body` should be of type `ArrayBuffer`, so we can't just use plain object, we need to convert it first.

This worked for me:
```typescript
  const toV2Url = (url: URL) => new URL(url.toString().replace('/v3'/, '/v2/'));

  const jsonToArrayBuffer = (json: Record<string, unknown>) => {
    const enc = new TextEncoder();
    return enc.encode(JSON.stringify(json)).buffer;
  };

  rest.patch(
    '*/construction/rfis-qa/v3/projects/:projectId/rfi-types/:rfiTypeId',
    async (req, res, ctx) => {
      const v3Body = await req.json();
      const v2Body = adapters.rfiTypes.patch.v3.to.v2(v3Body);

      const proxyReq = new MockedRequest(toV2Url(req.url), {
        ...req,
        body: jsonToArrayBuffer(v2Body),
      });

      const originalResponse = await ctx.fetch(proxyReq);
      const originalResponseData = await originalResponse.json();

      const responseV3 = adapters.rfiTypes.post.v2.to.v3(originalResponseData);

      return res(ctx.json(responseV3));
    },
  ),


```
